src/inmanta/server/services/paramservice.py

```diff
@@ -364,6 +364,79 @@ async def get_fact(self, env: data.Environment, rid: ResourceIdStr, id: uuid.UUI
             raise NotFound(f"Fact with id {id} does not exist")
         return param.as_fact()
 
+    async def _update_and_recompile(
+        self,
+        env: data.Environment,
+        name: str,
+        value: str,
+        source: ParameterSource,
+        metadata: dict[str, str],
+        recompile: bool,
+        resource_id: Optional[str] = None,
+        expires: Optional[bool] = None,
+    ) -> tuple[data.Parameter, Optional[list[str]]]:
+        """
+        Update a parameter or fact and optionally trigger recompilation.
+        """
+        # Validate parameter or fact
+        self._validate_parameter(name, resource_id, expires)
+
+        # Update parameter with new value and metadata
+        result = await self._update_param(env, name, value, source, resource_id, metadata, recompile, expires)
+        warnings = None
+        if result:
+            compile_metadata = {
+                "message": "Recompile model because one or more parameters were updated",
+                "type": "param",
+                "params": [(name, resource_id)],
+            }
+            warnings = await self.server_slice._async_recompile(env, False, metadata=compile_metadata)
+
+        # Retrieve the updated parameter/fact
+        param = await data.Parameter.get_list(environment=env.id, name=name)
```

Comment:
* The primary key is (environment, name, resource_id). So you can have an environment with several parameters with the same each belonging to a different resource. So you have to specify the resource_id to make sure we return the correct parameter.
* You also need to make sure that `resource_id=None` is converted into `resource_id=""`, because that's how we store them into the database.
* The `get_list()` can also become a `get_one()` like this.

```suggestion
        param = await data.Parameter.get_list(environment=env.id, name=name, resource_id=resource_id)
```